src/modules/job-manager/plugins/dependency-singleton.c

```diff
@@ -0,0 +1,344 @@
+ * (c.f. AUTHORS, NOTICE.LLNS, COPYING)
+ *
+ * This file is part of the Flux resource manager framework.
+ * For details, see https://github.com/flux-framework.
+ *
+ * SPDX-License-Identifier: LGPL-3.0
+\************************************************************/
+
+/* dependency-singleton.c
+ *
+ * Support singleton dependency scheme which places a dependency
+ * on a job which is only released when there are no other active
+ * jobs of the same userid and job name which are not also already
+ * held with a singleton dependency.
+ *
+ * Notes:
+ * - counts of active jobs with the same userid/job-name are maintained
+ *   in a global hash
+ * - jobs without an explicit name are ignored
+ * - jobs submitted with dependency=singleton are placed on a list, also
+ *   hashed by userid/job-name
+ * - when the active job count for a userid/job-name is decremented and
+ *   equals the count of singletons for that same pair, a singleton job
+ *   is released and removed from the list
+ * - it is an error to submit a singleton job without an explicit job name
+ *
+ */
+
+#if HAVE_CONFIG_H
+#include "config.h"
+#endif
+#include <jansson.h>
+#include <flux/core.h>
+#include <flux/jobtap.h>
+
+#include "src/common/libjob/idf58.h"
+#include "src/common/libczmqcontainers/czmq_containers.h"
+#include "src/common/libutil/hola.h"
+#include "ccan/str/str.h"
+#include "ccan/ptrint/ptrint.h"
+
+struct singleton_ctx {
+    zhashx_t *counts;
+    struct hola *singletons;
+};
+
+struct singleton {
+    flux_jobid_t id;
+};
+
+static struct singleton_ctx *global_ctx = NULL;
+
+static void singleton_ctx_destroy (struct singleton_ctx *sctx)
+{
+    if (sctx) {
+        int saved_errno = errno;
+        zhashx_destroy (&sctx->counts);
+        hola_destroy (sctx->singletons);
+        free (sctx);
+        errno = saved_errno;
+    }
+}
+
+static struct singleton_ctx *singleton_ctx_create (void)
+{
+    struct singleton_ctx *sctx;
+    int flags = HOLA_AUTOCREATE | HOLA_AUTODESTROY;
+
+    if (!(sctx = malloc (sizeof (*sctx)))
+        || !(sctx->counts = zhashx_new ())
+        || !(sctx->singletons = hola_create (flags)))
+        goto error;
+    return sctx;
+error:
+    singleton_ctx_destroy (sctx);
+    return NULL;
+}
+
+static int singleton_key_create (char *key,
+                                 size_t len,
+                                 uint32_t userid,
+                                 const char *name)
+{
+    if (snprintf (key, len, "%u:%s", userid, name) >= len)
+        return -1;
+    return 0;
+}
+
+static int singleton_list_push (flux_plugin_t *p,
+                                struct singleton_ctx *sctx,
+                                const char *key,
+                                flux_jobid_t id)
+{
+    struct singleton *s;
+
+    if (!(s = calloc (1, sizeof (*s)))
+        || flux_jobtap_job_aux_set (p, id, NULL, s, (flux_free_f) free) < 0) {
+        free (s);
+        return -1;
+    }
```

Comment:
Deferring destruction of these tiny objects until plugin unload is probably fine but I had to go look at the aux container implementation to remind myself that if it accumulates thousands of anonymous entries over the long runtime of a system instance, it doesn't slow down named lookups (b/c we put anons at the end - yay).

It might be nicer to just set the list destructor and also do an explicit free where these are popped off the list in the inactive callback.

Reply:
Perhaps it was missed that this is `flux_jobtap_job_aux_set()` not `flux_plugin_aux_set()`?

Reply:
OH so it's just the life of the job, well double nevermind then!  Sorry.